File: jscomp/test/arith_syntax.js
'use strict';

var Pervasives = require("../../lib/js/pervasives.js");

function str(e) {
  switch (e.TAG) {
    case "Numeral" :
        return Pervasives.string_of_float(e._0);
    case "Plus" :
        return str(e._0) + ("+" + str(e._1));
    case "Minus" :
        return str(e._0) + ("-" + str(e._1));
    case "Times" :
        return str(e._0) + ("*" + str(e._1));
    case "Divide" :
        return str(e._0) + ("/" + str(e._1));
    case "Negate" :
        return "-" + str(e._0);
    case "Variable" :
        return e._0;
    
  }
}

exports.str = str;
/* No side effect */

File: jscomp/test/arith_syntax.ml
type expression =
  | Numeral of float  (** non-negative integer constant *)
  | Plus of expression * expression  (** Addition [e1 + e2] *)
  | Minus of expression * expression  (** Difference [e1 - e2] *)
  | Times of expression * expression  (** Product [e1 * e2] *)
  | Divide of expression * expression  (** Quotient [e1 / e2] *)
  | Negate of expression  (** Opposite value [-e] *)
  | Variable of string

let rec str e =
  match e with
  | Numeral f -> string_of_float f
  | Plus (a, b) -> str a ^ "+" ^ str b
  | Minus (a, b) -> str a ^ "-" ^ str b
  | Times (a, b) -> str a ^ "*" ^ str b
  | Divide (a, b) -> str a ^ "/" ^ str b
  | Negate a -> "-" ^ str a
  | Variable s -> s
